File: app/src/main/java/cielo/ordermanager/sdk/sample/CancelPaymentActivity.java
package cielo.ordermanager.sdk.sample;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cielo.ordermanager.sdk.adapter.PaymentRecyclerViewAdapter;
import cielo.ordermanager.sdk.listener.RecyclerItemClickListener;
import cielo.orders.domain.Credentials;
import cielo.orders.domain.Order;
import cielo.sdk.order.OrderManager;
import cielo.sdk.order.ServiceBindListener;
import cielo.sdk.order.cancellation.CancellationListener;
import cielo.sdk.order.payment.Payment;
import cielo.sdk.order.payment.PaymentError;

public class CancelPaymentActivity extends AppCompatActivity {

    private final String TAG = "CANCEL_PAYMENT";

    private Order order;
    private Payment payment;
    private OrderManager orderManager;
    private boolean isBound = false;

    @BindView(R.id.btn_cancel_payment)
    Button cancelButton;

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_cancel_payment);
        ButterKnife.bind(this);

        cancelButton.setEnabled(false);

        configSDK();

        this.order = (Order) getIntent().getSerializableExtra("SELECTED_ORDER");

        if (order != null && !order.getPayments().isEmpty()) {

            final List<Payment> paymentList = order.getPayments();

            recyclerView.setLayoutManager(new LinearLayoutManager(CancelPaymentActivity.this));
            recyclerView.setAdapter(
                    new PaymentRecyclerViewAdapter(paymentList));

            Log.i(TAG, "payments: " + paymentList);
            for (Payment pay : paymentList) {
                Log.i("Payment: ", pay.getExternalId() + " - " + pay.getAmount());
            }

            recyclerView.addOnItemTouchListener(
                    new RecyclerItemClickListener(CancelPaymentActivity.this, recyclerView, new RecyclerItemClickListener.OnItemClickListener() {
                        @Override
                        public void onItemClick(View view, int position) {

                            payment = paymentList.get(position);
                            cancelButton.setEnabled(true);

                        }

                        @Override
                        public void onLongItemClick(View view, int position) {
                        }
                    })
            );

        } else {


        }
    }

    @OnClick(R.id.btn_cancel_payment)
    public void cancelPayment() {

        if (order != null && order.getPayments().size() > 0) {

            orderManager.cancelOrder(this, order.getId(), payment, new CancellationListener() {
                @Override
                public void onSuccess(Order cancelledOrder) {
                    Log.d(TAG, "ON SUCCESS");
                    cancelledOrder.cancel();
                    orderManager.updateOrder(cancelledOrder);

                    order = cancelledOrder;
                    resetUI();
                }

                @Override
                public void onCancel() {
                    Log.d(TAG, "ON CANCEL");
                    resetUI();
                }

                @Override
                public void onError(PaymentError paymentError) {
                    Log.d(TAG, "ON ERROR");
                    resetUI();
                }
            });
        } else {

        }
    }

    private void resetUI() {
        payment = null;
        cancelButton.setEnabled(false);
    }

    private void configSDK() {
        Credentials credentials = new Credentials("1234", "1234");
        orderManager = new OrderManager(credentials, this);
        orderManager.bind(this, new ServiceBindListener() {
            @Override
            public void onServiceBound() {
                isBound = true;
            }

            @Override
            public void onServiceUnbound() {
                isBound = false;
            }
        });
    }

}

File: app/src/main/java/com/cielo/ordermanager/sdk/sample/activities/CancelPaymentActivity.kt
package com.cielo.ordermanager.sdk.sample.activities

import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cielo.orders.domain.CancellationRequest
import cielo.orders.domain.Order
import cielo.sdk.order.cancellation.CancellationListener
import cielo.sdk.order.payment.Payment
import cielo.sdk.order.payment.PaymentError
import com.cielo.ordermanager.sdk.databinding.ActivityCancelPaymentBinding
import com.cielo.ordermanager.sdk.listener.RecyclerItemClickListener
import com.cielo.ordermanager.sdk.orders.OrderManagerController
import com.cielo.ordermanager.sdk.sample.adapter.PaymentRecyclerViewAdapter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class CancelPaymentActivity : AppCompatActivity() {

    private val TAG = "CANCEL_PAYMENT"
    private lateinit var order: Order
    private var payment: Payment? = null
    private val orderManagerController: OrderManagerController = OrderManagerController.getInstance(this)
    private lateinit var binding: ActivityCancelPaymentBinding
    private val scope = CoroutineScope(Dispatchers.Default)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCancelPaymentBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.btnCancelPayment.isEnabled = false


        order = intent.getSerializableExtra("SELECTED_ORDER") as Order

        if (order.payments.isNotEmpty()) {
            val paymentList = order.payments

            binding.recyclerView.layoutManager = LinearLayoutManager(this)
            binding.recyclerView.adapter = PaymentRecyclerViewAdapter(paymentList)

            Log.i(TAG, "payments: $paymentList")

            for (pay in paymentList) {
                Log.i("Payment: ", "${pay.externalId} - ${pay.amount}")
            }

            binding.recyclerView.addOnItemTouchListener(
                RecyclerItemClickListener(this, binding.recyclerView, object : RecyclerItemClickListener.OnItemClickListener {
                    override fun onItemClick(view: View, position: Int) {
                        payment = paymentList[position]
                        binding.btnCancelPayment.isEnabled = true
                    }

                    override fun onLongItemClick(view: View, position: Int) {
                        // Implementação do long click, se necessário
                    }
                })
            )
        }

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Dados para Cancelamento"

        binding.btnCancelPayment.setOnClickListener {
            scope.launch {
                cancelPayment()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        finish()
        return super.onOptionsItemSelected(item)
    }

    private suspend fun cancelPayment() {
        if (order.payments.isNotEmpty()) {
            payment?.let {
                val request = CancellationRequest.Builder()
                    .orderId(order.id)
                    .authCode(it.authCode)
                    .cieloCode(it.cieloCode)
                    .value(it.amount)
                    .ec(it.merchantCode)
                    .build()

                orderManagerController.cancelOrder(request, object : CancellationListener {
                    override fun onSuccess(cancelledOrder: Order) {
                        Log.d(TAG, "ON SUCCESS")
                        cancelledOrder.cancel()
                        orderManagerController.updateOrder(cancelledOrder)

                        Toast.makeText(this@CancelPaymentActivity, "SUCESSO", Toast.LENGTH_SHORT).show()
                        order = cancelledOrder
                        resetUI()
                    }

                    override fun onCancel() {
                        Log.d(TAG, "ON CANCEL")
                        Toast.makeText(this@CancelPaymentActivity, "CANCELADO", Toast.LENGTH_SHORT).show()
                        resetUI()
                    }

                    override fun onError(paymentError: PaymentError) {
                        Log.d(TAG, "ON ERROR")
                        Toast.makeText(this@CancelPaymentActivity, "ERRO", Toast.LENGTH_SHORT).show()
                        resetUI()
                    }
                })
            } ?: run {
                Toast.makeText(this, "Pagamento não pode ser null", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun resetUI() {
        payment = null
        binding.btnCancelPayment.isEnabled = false
    }
}